crates/circuit/src/packed_instruction.rs

```diff
@@ -0,0 +1,497 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+#[cfg(feature = "cache_pygates")]
+use std::cell::RefCell;
+use std::ptr::NonNull;
+
+use pyo3::intern;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+
+use smallvec::SmallVec;
+
+use crate::circuit_instruction::ExtraInstructionAttributes;
+use crate::imports::DEEPCOPY;
+use crate::operations::{OperationRef, Param, PyGate, PyInstruction, PyOperation, StandardGate};
+
+/// The logical discriminant of `PackedOperation`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(u8)]
+enum PackedOperationType {
+    // It's important that the `StandardGate` item is 0, so that zeroing out a `PackedOperation`
+    // will make it appear as a standard gate, which will never allow accidental dangling-pointer
+    // dereferencing.
+    StandardGate = 0,
+    Gate = 1,
+    Instruction = 2,
+    Operation = 3,
+}
+unsafe impl ::bytemuck::CheckedBitPattern for PackedOperationType {
+    type Bits = u8;
+
+    fn is_valid_bit_pattern(bits: &Self::Bits) -> bool {
+        *bits < 4
+    }
+}
+unsafe impl ::bytemuck::NoUninit for PackedOperationType {}
+
+/// A bit-packed `OperationType` enumeration.
+///
+/// This is logically equivalent to:
+///
+/// ```rust
+/// enum Operation {
+///     Standard(StandardGate),
+///     Gate(Box<PyGate>),
+///     Instruction(Box<PyInstruction>),
+///     Operation(Box<PyOperation>),
+/// }
+/// ```
+///
+/// including all ownership semantics, except it bit-packs the enumeration into a single pointer.
+/// This works because `PyGate` (and friends) have an alignment of 8, so pointers to them always
+/// have the low three bits set to 0, and `StandardGate` has a width much smaller than a pointer.
+/// This lets us store the enum discriminant in the low data bits, and then type-pun a suitable
+/// bitmask on the contained value back into proper data.
+///
+/// Explicity, this is logical memory layout of `PackedOperation` on a 64-bit system, written out as
```

Comment:
"Explicitly"

Reply:
Done in f991157574cfd9c462418366486b307e712f9059.